File: Antidote/QRViewerController.m
//
//  QRViewerController.m
//  Antidote
//
//  Created by Dmitry Vorobyov on 19.07.14.
//  Copyright (c) 2014 dvor. All rights reserved.
//

#import <CoreImage/CoreImage.h>

#import "QRViewerController.h"
#import "CopyLabel.h"
#import "NSString+Utilities.h"

@interface QRViewerController ()

@property (strong, nonatomic) UIButton *closeButton;
@property (strong, nonatomic) UIImageView *imageView;
@property (strong, nonatomic) CopyLabel *label;

@property (strong, nonatomic) NSString *text;

@end

@implementation QRViewerController

- (instancetype)initWithText:(NSString *)text
{
    self = [super init];

    if (self) {
        self.text = text;
    }

    return self;
}

- (void)loadView
{
    CGRect frame = CGRectZero;
    frame.size = [[UIScreen mainScreen] applicationFrame].size;

    self.view = [[UIView alloc] initWithFrame:frame];
    self.view.backgroundColor = [UIColor whiteColor];

    [self createCloseButton];
    [self createImageView];
    [self createLabel];
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];

    [self adjustSubviews];
}

- (void)viewDidLoad
{
    [super viewDidLoad];

    CIFilter *filter = [CIFilter filterWithName:@"CIQRCodeGenerator"];
    [filter setDefaults];
    [filter setValue:[self.text dataUsingEncoding:NSUTF8StringEncoding]
              forKey:@"inputMessage"];

    CIImage *outputImage = [filter outputImage];

    CIContext *context = [CIContext contextWithOptions:nil];
    CGImageRef cgImage = [context createCGImage:outputImage fromRect:outputImage.extent];

    self.imageView.image = [UIImage imageWithCGImage:cgImage scale:1.0 orientation:UIImageOrientationUp];

    CGImageRelease(cgImage);
}

#pragma mark -  Actions

- (void)closeButtonPressed
{
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark -  Private

- (void)createCloseButton
{
    self.closeButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.closeButton setTitle:NSLocalizedString(@"Close", @"QRViewerController") forState:UIControlStateNormal];
    [self.closeButton addTarget:self
                         action:@selector(closeButtonPressed)
               forControlEvents:UIControlEventTouchUpInside];

    [self.view addSubview:self.closeButton];
}

- (void)createImageView
{
    self.imageView = [UIImageView new];
    [self.view addSubview:self.imageView];
}

- (void)createLabel
{
    self.label = [CopyLabel new];
    self.label.textColor = [UIColor grayColor];
    self.label.backgroundColor = [UIColor clearColor];
    self.label.numberOfLines = 0;
    self.label.text = self.text;
    [self.view addSubview:self.label];
}

- (void)adjustSubviews
{
    CGRect frame = CGRectZero;

    {
        NSString *title = [self.closeButton titleForState:UIControlStateNormal];
        UIFont *font = self.closeButton.titleLabel.font;

        frame = CGRectZero;
        frame.size = [title stringSizeWithFont:font];
        frame.origin.x = self.view.bounds.size.width - frame.size.width - 15.0;
        frame.origin.y = 30.0;
        self.closeButton.frame = frame;
    }

    {
        frame = CGRectZero;
        frame.size.width = frame.size.height = self.view.bounds.size.width;
        frame.origin.y = (self.view.bounds.size.height - frame.size.height) / 2;
        self.imageView.frame = frame;

        self.imageView.image = [self resizeImage:self.imageView.image newSize:frame.size];
    }

    {
        CGFloat xIndentation = 10.0;
        CGFloat maxWidth = self.view.bounds.size.width - 2 * xIndentation;

        frame = CGRectZero;
        frame.size = [self.label.text stringSizeWithFont:self.label.font
                                       constrainedToSize:CGSizeMake(maxWidth, CGFLOAT_MAX)];
        frame.origin.x = xIndentation;
        frame.origin.y = CGRectGetMaxY(self.imageView.frame) + 10.0;
        self.label.frame = frame;
    }
}

- (UIImage *)resizeImage:(UIImage *)image newSize:(CGSize)newSize
{
    UIImage *resized = nil;

    UIGraphicsBeginImageContext(newSize);
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetInterpolationQuality(context, kCGInterpolationNone);
    [image drawInRect:CGRectMake(0, 0, newSize.width, newSize.height)];
    resized = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();

    return resized;
}

@end

File: Antidote/QRViewerController.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import UIKit
import SnapKit

protocol QRViewerControllerDelegate: class {
    func qrViewerControllerDidFinishPresenting()
}

class QRViewerController: UIViewController {
    weak var delegate: QRViewerControllerDelegate?

    fileprivate let theme: Theme
    fileprivate let text: String

    fileprivate var previousBrightness: CGFloat = 1.0

    fileprivate var closeButton: UIButton!
    fileprivate var imageView: UIImageView!

    init(theme: Theme, text: String) {
        self.theme = theme
        self.text = text

        super.init(nibName: nil, bundle: nil)

        edgesForExtendedLayout = UIRectEdge()
    }

    required convenience init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func loadView() {
        loadViewWithBackgroundColor(theme.colorForType(.NormalBackground))

        installNavigationItems()
        createViews()
        installConstraints()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        previousBrightness = UIScreen.main.brightness
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

        UIScreen.main.brightness = previousBrightness
    }
}

extension QRViewerController {
    func closeButtonPressed() {
        delegate?.qrViewerControllerDidFinishPresenting()
    }
}

private extension QRViewerController {
    func installNavigationItems() {
        navigationItem.rightBarButtonItem = UIBarButtonItem(
                barButtonSystemItem: .done,
                target: self,
                action: #selector(QRViewerController.closeButtonPressed))
    }

    func createViews() {
        imageView = UIImageView(image: qrImageFromText())
        imageView.contentMode = .scaleAspectFit
        view.addSubview(imageView)
    }

    func installConstraints() {
        imageView.snp.makeConstraints {
            $0.center.equalTo(view)
            $0.width.lessThanOrEqualTo(view.snp.width)
            $0.width.lessThanOrEqualTo(view.snp.height)
            $0.width.equalTo(imageView.snp.height)
        }
    }

    func qrImageFromText() -> UIImage {
        let filter = CIFilter(name:"CIQRCodeGenerator")!
        filter.setDefaults()
        filter.setValue(text.data(using: String.Encoding.utf8), forKey: "inputMessage")

        let ciImage = filter.outputImage!
        let screenBounds = UIScreen.main.bounds

        let scale = min(screenBounds.size.width / ciImage.extent.size.width, screenBounds.size.height / ciImage.extent.size.height)
        let transformedImage = ciImage.applying(CGAffineTransform(scaleX: scale, y: scale))

        return UIImage(ciImage: transformedImage)
    }
}
